wgengine/netstack/netstack.go

```diff
@@ -381,18 +382,71 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 		ns.logf("Could not create endpoint, exiting")
 		return
 	}
+	localAddr, err := ep.GetLocalAddress()
+	if err != nil {
+		return
+	}
+	remoteAddr, err := ep.GetRemoteAddress()
+	if err != nil {
+		return
+	}
 	c := gonet.NewUDPConn(ns.ipstack, &wq, ep)
-	go echoUDP(c)
+	go ns.forwardUDP(c, &wq, localAddr, remoteAddr)
 }
 
-func echoUDP(c *gonet.UDPConn) {
-	buf := make([]byte, 1500)
-	for {
-		n, err := c.Read(buf)
+func (ns *Impl) forwardUDP(client *gonet.UDPConn, wq *waiter.Queue, clientLocalAddr, clientRemoteAddr tcpip.FullAddress) {
+	port := clientLocalAddr.Port
+	ns.logf("[v2] netstack: forwarding incoming UDP connection on port %v", port)
+	serverLocalAddr := &net.UDPAddr{Port: int(clientRemoteAddr.Port)}
+	serverRemoteAddr := &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: int(port)}
+	server, err := net.ListenUDP("udp", serverLocalAddr)
+	if err != nil {
+		ns.logf("netstack: could not bind local port %v: %v, trying again with random port", clientRemoteAddr.Port, err)
+		server, err = net.ListenUDP("udp", nil)
 		if err != nil {
-			break
+			ns.logf("netstack: could not connect to local UDP server on port %v: %v", port, err)
+			return
 		}
-		c.Write(buf[:n])
 	}
-	c.Close()
+	var cancels [2]context.CancelFunc
+	timer := time.AfterFunc(2*time.Minute, func() {
+		for _, c := range cancels {
+			if c != nil {
+				c()
+			}
+		}
+		ns.logf("netstack: forwarder UDP connection on port %v closed", port)
```

Comment:
I'd log first, then close things.

also, you need to close `server` (which I'd name `backendConn`).

and the `client *gonet.UDPConn` isn't being closed.